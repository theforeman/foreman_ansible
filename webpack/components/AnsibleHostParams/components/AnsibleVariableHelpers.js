import React from 'react';
import { find, lowerCase } from 'lodash';
import { sprintf, translate as __ } from 'foremanReact/common/I18n';

export const overrideFieldName = (lookupKey, attr, modelField) =>
  `${modelField}[lookup_values_attributes][${lookupKey.id}][${attr}]`;

export const updateFieldDisabled = (overriden, ommited) =>
  !overriden || ommited;

export const matcherElement = modelName =>
  modelName === 'hostgroup' ? 'hostgroup' : 'fqdn';

export const fieldElement = modelName =>
  modelName === 'hostgroup' ? 'hostgroup' : 'host';

export const constructId = (role, lookupKey) =>
  `ansible_role_${role.id}_params[${lookupKey.id}]`;

export const roleNameColumn = role => (
  <td className="elipsis" rowSpan={role.ansibleVariables.length}>
    {role.name}
  </td>
);

export const initLookupValue = (lookupKey, formObject) => {
  const modelName = lowerCase(formObject.resourceName);
  const valueMatch = matcherElement(modelName);

  if (!lookupKey.currentOverride) {
    return {
      element: valueMatch,
      value: lookupKey.defaultValue,
      omit: false,
      overriden: false,
      defaultValue: lookupKey.defaultValue,
    };
  }

  if (lookupKey.currentOverride.element === valueMatch) {
    const found = find(
      lookupKey.overrideValues,
      value =>
        value.match === `${valueMatch}=${lookupKey.currentOverride.elementName}`
    );

    const override = {
      ...lookupKey.currentOverride,
      id: found.id,
      omit: found.omit,
      overriden: true,
      defaultValue: lookupKey.defaultValue,
    };

    return override;
  }

  return { ...lookupKey.currentOverride, overriden: false, omit: false };
};

export const sanitizeFormValue = val => (val === null ? '' : val);

export const sanitizeFieldValue = lookupValue =>
  sanitizeFormValue(
    lookupValue.value ? lookupValue.value : lookupValue.defaultValue
  );

export const shouldDestroy = (lookupValue, fieldDisabled, fieldOmmited) =>
  lookupValue.overriden && fieldDisabled && !fieldOmmited;

export const shouldAddFields = (lookupValue, fieldDisabled, fieldOmmited) =>
  !fieldDisabled || fieldOmmited || (lookupValue.overriden && fieldDisabled);

export const lookupKeyValidations = (
  lookupKey,
  fieldValue,
  fieldDisabled,
  resourceError,
  dirty
) => {
  const validRes = validOverrideFactory();
  const invalidText = `<b>${__('Please change!')}</b>`;

  const validateValue = (condition, onInvalid) => {
    if (condition(lookupKey.validatorRule, fieldValue)) {
      return validRes;
    }
    return onInvalid;
  };

  if (fieldDisabled) {
    return validRes;
  }

  if (resourceError && fieldValue && !dirty) {
    const errorText = Object.entries(resourceError.errors)
      .reduce((memo, [key, value]) => memo.concat([value]), [])
      .join(', ');
    return invalidOverrideFactory(invalidText, errorText);
  }

  if (lookupKey.required) {
    const pleaseChange = `${__(
      'Required parameter with invalid value.'
    )}<br/><b>${__('Please change!')}</b><br/>`;

    switch (lookupKey.validatorType) {
      case 'None':
        return validateValue(
          (rule, value) => value,
          invalidOverrideFactory(
            `${__('Required parameter without value.')}<br/><b>${__(
              'Please override!'
            )}</b><br/>`,
            __("Value can't be blank")
          )
        );
      case 'regexp': {
        return validateValue(
          (rule, value) => new RegExp(rule).test(value),
          invalidOverrideFactory(
            pleaseChange,
            sprintf(
              __('Invalid value, expected to match a regex: %s'),
              lookupKey.validatorRule
            )
          )
        );
      }
      case 'list':
        return validateValue(
          (rule, value) => rule.split(',').find(item => item.trim() === value),
          invalidOverrideFactory(
            pleaseChange,
            sprintf(
              __('Invalid value, expected one of: %s'),
              lookupKey.validatorRule
            )
          )
        );
      default:
        break;
    }
  }

  if (fieldValue) {
    return validRes;
  }

  return {
    text: `${__('Optional parameter without value.')}<br/><i>${__(
      'Still managed by Foreman, the value will be empty.'
    )}</i><br/>`,
    icon: 'warning-triangle-o',
    valid: true,
  };
};

export const invalidOverrideFactory = (text, msg) => ({
  text,
  icon: 'error-circle-o',
  valid: false,
  msg,
});

export const validOverrideFactory = () => ({
  text: '',
  icon: 'info',
  valid: true,
});

export const showLookupValue = (hidden, lookupKey) => {
  if (hidden) {
    return '*****';
  }

  if (lookupKey.currentOverride) {
    return lookupKey.currentOverride.value;
  }

  return lookupKey.defaultValue;
};

export const formatMatcher = currentOverride =>
  currentOverride
    ? `${currentOverride.element} (${currentOverride.elementName})`
    : 'Default value';
