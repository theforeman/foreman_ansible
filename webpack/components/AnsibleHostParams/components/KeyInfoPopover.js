import React from 'react';
import PropTypes from 'prop-types';
import { Popover, Icon } from 'patternfly-react';

import GenericPopover from './GenericPopover';

import { showLookupValue, formatMatcher } from './AnsibleVariableHelpers';

const KeyInfoPopover = ({ lookupKey, hidden, keyErrors }) => {
  const button = (
    <a href="#" className="popover-pf-info">
      <Icon type="pf" name={keyErrors.icon} />
    </a>
  );

  const keyOverlay = (key, isHidden, text) => (
    <Popover id="popover" title="Original value info">
      <div>
        {text && <span dangerouslySetInnerHTML={{ __html: text }} />}
        <b>Description: </b>
        {key.description}
        <br />
        <b>Type: </b> {key.parameterType}
        <br />
        <b>Matcher: </b>
        {formatMatcher(key.currentOverride)}
        <br />
        <b>Inherited Value: </b>
        {showLookupValue(isHidden, key)}
        <br />
      </div>
    </Popover>
  );

  return (
    <div style={{ textAlign: 'center' }}>
      <GenericPopover
        popoverOverlay={keyOverlay(lookupKey, hidden, keyErrors.text)}
        button={button}
      />
    </div>
  );
};

KeyInfoPopover.propTypes = {
  lookupKey: PropTypes.object.isRequired,
  hidden: PropTypes.bool.isRequired,
  keyErrors: PropTypes.object.isRequired,
};

export default KeyInfoPopover;
