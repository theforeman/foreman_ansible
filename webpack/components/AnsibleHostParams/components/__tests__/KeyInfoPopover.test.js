import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import KeyInfoPopover from '../KeyInfoPopover';

import { validOverrideFactory } from '../AnsibleVariableHelpers';
import {
  ellipseLookupKey,
  ellipseLookupValue,
  squareLookupKey,
  squareLookupValue,
} from '../../AnsibleHostParams.fixtures';

const fixtures = {
  'should render when value is hidden': {
    keyErrors: validOverrideFactory(),
    lookupKey: ellipseLookupKey,
    lookupValue: ellipseLookupValue,
    hidden: true,
  },
  'should render when value is not hidden': {
    keyErrors: validOverrideFactory(),
    lookupKey: squareLookupKey,
    lookupValue: squareLookupValue,
    hidden: false,
  },
};

describe('KeyInfoPopover', () =>
  testComponentSnapshotsWithFixtures(KeyInfoPopover, fixtures));
