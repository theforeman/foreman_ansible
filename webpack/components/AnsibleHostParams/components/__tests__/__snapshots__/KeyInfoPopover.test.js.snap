// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyInfoPopover should render when value is hidden 1`] = `
<div
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <GenericPopover
    button={
      <a
        className="popover-pf-info"
        href="#"
      >
        <Icon
          name="info"
          type="pf"
        />
      </a>
    }
    placement="top"
    popoverId={null}
    popoverOverlay={
      <Popover
        bsClass="popover"
        id="popover"
        placement="right"
        title="Original value info"
      >
        <div>
          
          <b>
            Description: 
          </b>
          
          <br />
          <b>
            Type: 
          </b>
           
          boolean
          <br />
          <b>
            Matcher: 
          </b>
          Default value
          <br />
          <b>
            Inherited Value: 
          </b>
          *****
          <br />
        </div>
      </Popover>
    }
    popoverText={null}
    popoverTitle={null}
    trigger="click"
  />
</div>
`;

exports[`KeyInfoPopover should render when value is not hidden 1`] = `
<div
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <GenericPopover
    button={
      <a
        className="popover-pf-info"
        href="#"
      >
        <Icon
          name="info"
          type="pf"
        />
      </a>
    }
    placement="top"
    popoverId={null}
    popoverOverlay={
      <Popover
        bsClass="popover"
        id="popover"
        placement="right"
        title="Original value info"
      >
        <div>
          
          <b>
            Description: 
          </b>
          Square param
          <br />
          <b>
            Type: 
          </b>
           
          boolean
          <br />
          <b>
            Matcher: 
          </b>
          hostgroup (first hostgroup)
          <br />
          <b>
            Inherited Value: 
          </b>
          <br />
        </div>
      </Popover>
    }
    popoverText={null}
    popoverTitle={null}
    trigger="click"
  />
</div>
`;
